### PR TITLE
Media: Implement picker mode

### PR DESCRIPTION
**Pull request by kean**
Part of https://github.com/wordpress-mobile/WordPress-iOS/issues/21457

- Rename `MediaViewController` to `SiteMediaViewController` (more clear what it's for)
- Extract `SiteMediaCollectionViewController` to manage the media collection and use it in a new `SiteMediaViewController`
- Add `SiteMediaPickerViewController`
- `SiteMediaViewController` now displays the number of selected items in the toolbar

I've been writing ViewModels with reactive programming for the past five years, so I'm going old-school here with full MVC 😄 I think it's easier to understand when `NSFetchResultsController` is involved.

media picker

## To test:

We don't have any particular requirements for this screen, and it's a bit early to test until we get to phase 2. But if you want to see it, add this somewhere:

// Single selection

```swift
        DispatchQueue.main.asyncAfter(deadline: .now() + .seconds(2)) {
            let vc = SiteMediaPickerViewController(blog: self.blog!)
            let nav = UINavigationController(rootViewController: vc)
            self.present(nav, animated: true)
        }
```

// Video with multiple selection

```swift
        DispatchQueue.main.asyncAfter(deadline: .now() + .seconds(2)) {
            let vc = SiteMediaPickerViewController(blog: self.blog!, filter: [.video], allowsMultipleSelection: true)
            let nav = UINavigationController(rootViewController: vc)
            self.present(nav, animated: true)
        }
```

- Verify that the filter works
- Verify that search works
- Verify that selection works and updates the state of the "Done" button
- Verify that single selection works (nothing will happen; it calls the delegate)
- Verify that "Cancel" works (nothing will happen; it calls the delegate)
- Verify that it doesn't show pending uploads

## Regression Notes
1. Potential unintended areas of impact: none
2. What I did to test those areas of impact (or what existing automated tests I relied on): manual
3. What automated tests I added (or what prevented me from doing so): none

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

UI Changes testing checklist:
- [ ] Portrait and landscape orientations.
- [x] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] VoiceOver.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] iPhone and iPad. 
- [ ] Multi-tasking: Split view and Slide over. (iPad)

### CODE REVIEW

**Comment by kean** [comment]
Hey, @osullivanchris. We haven't talked about the Site Media Picker yet, but it also needs to change as it's based on the screen we are updating. I implemented the initial version with native components.

Notable changes:

- Unlike the Site Media screen, the search bar is always visible, even if you scroll down
- Doesn't show unsynced assets (matches the existing behavior, but we could probably still show them, just no allow to select until the upload completes)
- No editing (matches the existing behavior)
- "Done" is in the navigation bar, just like in the native picker. I think we should keep it consistent.

I think I could add preview no long-tap or something like that. Apple also has a "Show Selection" button that opens the selected items fullscreen.

File: WordPress/Classes/ViewRelated/Media/SiteMedia/SiteMediaViewController.swift [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+        let image = UIImage(systemName: "plus", withConfiguration: config) ?? .gridicon(.plus)
+        button.setImage(image, for: .normal)
+        button.addAction(UIAction { [weak self] _ in
+            QuickStartTourGuide.shared.visited(.mediaUpload)
+            self?.buttonAddMedia.shouldShowSpotlight = false
+        }, for: .menuActionTriggered)
+        button.menu = buttonAddMediaMenuController.makeMenu(for: self)
+        button.showsMenuAsPrimaryAction = true
+        button.accessibilityLabel = NSLocalizedString("Add", comment: "Accessibility label for add button to add items to the user's media library")
+        button.accessibilityHint = NSLocalizedString("Add new media", comment: "Accessibility hint for add button to add items to the user's media library")
```
Comment (momo-ozawa) [inline WordPress/Classes/ViewRelated/Media/SiteMedia/SiteMediaViewController.swift]: These strings are missing reverse dns keys

**Review by momo-ozawa — COMMENTED** [review]
@kean Multiple selection works as described, but I'm not able to select a photo for single selection

**Comment by kean** [comment]
> @kean Multiple selection works as described, but I'm not able to select a photo for single selection

Visually, nothing happens. It calls a delegate, and the delegate will be responsible for dismissing it.

File: WordPress/Classes/ViewRelated/Media/SiteMedia/SiteMediaViewController.swift (hunk 1 above)
Comment (kean) [reply WordPress/Classes/ViewRelated/Media/SiteMedia/SiteMediaViewController.swift]: Fixed!
I copied it from the original implementation 😄

**Review by momo-ozawa — APPROVED** [review]

**Pull request merged**